rslib/src/collection/backup.rs

```diff
@@ -0,0 +1,348 @@
+            .map(|datetime| Self {
+                path: entry.path(),
+                datetime,
+            })
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct BackupThinner {
+    yesterday: i32,
+    last_kept_day: i32,
+    last_kept_week: i32,
+    last_kept_month: u32,
+    limits: BackupLimits,
+}
+
+#[derive(Debug, Clone, Copy, PartialEq)]
+enum BackupStage {
+    Daily,
+    Weekly,
+    Monthly,
+}
+
+impl BackupThinner {
+    fn new(limits: BackupLimits) -> Self {
+        Self {
+            yesterday: Local::today().num_days_from_ce() - 1,
+            last_kept_day: i32::MAX,
+            last_kept_week: i32::MAX,
+            last_kept_month: u32::MAX,
+            limits,
+        }
+    }
+
+    fn thin(mut self, backups: impl Iterator<Item = Backup>) {
+        use BackupStage::*;
+        for backup in backups
+            .sorted_unstable_by_key(|b| b.datetime.timestamp())
+            .rev()
+        {
+            if self.is_recent(&backup) {
+                continue;
+            } else if self.remaining(Daily) {
+                self.keep_or_delete(Daily, &backup);
+            } else if self.remaining(Weekly) {
+                self.keep_or_delete(Weekly, &backup);
+            } else if self.remaining(Monthly) {
+                self.keep_or_delete(Monthly, &backup);
+            } else {
+                self.delete(&backup);
+            }
+        }
+    }
+
+    fn is_recent(self, backup: &Backup) -> bool {
+        backup.day() >= self.yesterday
+    }
+
+    fn remaining(self, stage: BackupStage) -> bool {
+        match stage {
+            BackupStage::Daily => self.limits.daily > 0,
+            BackupStage::Weekly => self.limits.weekly > 0,
+            BackupStage::Monthly => self.limits.monthly > 0,
+        }
+    }
+
+    fn keep_or_delete(&mut self, stage: BackupStage, backup: &Backup) {
+        let keep = match stage {
+            BackupStage::Daily => backup.day() < self.last_kept_day,
+            BackupStage::Weekly => backup.week() < self.last_kept_week,
+            BackupStage::Monthly => backup.month() < self.last_kept_month,
+        };
+        if keep {
+            self.keep(stage, backup);
+        } else {
+            self.delete(backup);
+        }
+    }
+
+    /// Adjusts limits as per the stage of the kept backup.
+    fn keep(&mut self, stage: BackupStage, backup: &Backup) {
+        match stage {
+            BackupStage::Daily => {
+                self.limits.daily -= 1;
+                self.last_kept_day = backup.day();
+            }
+            BackupStage::Weekly => {
+                self.limits.weekly -= 1;
+                self.last_kept_week = backup.week();
+            }
+            BackupStage::Monthly => {
+                self.limits.monthly -= 1;
+                self.last_kept_month = backup.month();
+            }
+        }
+    }
+
+    /// Tries to delete the [Backup], discarding any errors.
+    fn delete(&mut self, backup: &Backup) {
```

Comment:
It'd be nice to able to test this without writing any files. Should I make `BackupThinner::new` take a `dyn Fn(&Backup)`? Then in testing we could pass in some closure that allows to check which `Backup`s would get deleted.

Reply:
How about separating out the logic and the I/O? Eg we have a function that takes a list of backups and returns a list of ones that need deletion (or mutates a delete/keep flag on them). That's easy to test, and then the actual deletion is just a matter of looping over the selected files and removing them.

Reply:
Oh, that's even better. 🙂